### PR TITLE
[BPK-878] Add rn animate height

### PR DESCRIPTION
**Pull request by georgegillams**
Added animate-height component as it seemed sensible to make this separate to banner-alert (which will probably make use of it)

### CODE REVIEW

File: native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  animationDurationBase,
+} = tokens;
+
+class BpkAnimateHeight extends React.Component {
+
+  state = {
+    expanded: null,
+    expandedHeight: null,
+    collapsedHeight: 0,
+    height: null,
```
Comment (georgegillams) [inline native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js]: Should be `height: new Animated.Height(0),`

File: native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      <Animated.View
+        style={{
+          ...style,
+          overflow: 'hidden',
+          height,
+        }}
+        {...rest}
+      >
+        <View
+          onLayout={event => this.state.height === null ? this.setExpandedHeight(event, expanded) : undefined}
```
Comment (georgegillams) [inline native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js]: `this.state.heightSet ? ...`

File: native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  setExpandedHeight = (event, expandedImmediately) => {
+    // if (this.state.expandedHeightSet) return;
+    const { height } = event.nativeEvent.layout;
+    this.setState({
+      // if the component has started expanded, we should now set the height of the container
+      // why on earth does a value of 1 instead of 0 make this break!
+      height: new Animated.Value(expandedImmediately ? height : 0),
+      expandedHeight: height,
```
Comment (georgegillams) [inline native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js]: `heightSet: {true}`,

File: changelog.md [hunk 4]
```diff
@@ -17,8 +18,13 @@ _Nothing yet..._
 
 **Added:**
 - react-native-bpk-theming  0.0.1 => 1.0.0
+<<<<<<< HEAD
```
Comment (matteo-hertel) [inline changelog.md]: 😱

File: native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+          {this.props.children}
+        </View>
+      </Animated.View>
+    );
+  }
+}
+
+BpkAnimateHeight.propTypes = {
+  expanded: PropTypes.bool.isRequired,
+  children: PropTypes.oneOfType([
```
Comment (matteo-hertel) [inline native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js]: why not something like `children: PropTypes.node.isRequired,`? are you expecting an array of children?

File: native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight-test.common.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        >
+          {animateHeightContent}
+        </BpkAnimateHeight>).toJSON();
+      expect(tree).toMatchSnapshot();
+    });
+
+    it('should render correctly with n children', () => {
+      const tree = renderer.create(
+        <BpkAnimateHeight
+          expanded
```
Comment (matteo-hertel) [inline native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight-test.common.js]: `should render correctly with n children` but there are 0 children, is this correct? you should update either the description or the test itself

File: packages/bpk-tokens/src/common/base/animations.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "aliases": {
+    "ANIMATION_DURATION_XS": "50ms",
```
Comment (matteo-hertel) [inline packages/bpk-tokens/src/common/base/animations.json]: just a quick note on this, these values have been discussed or are just arbitrary values?

File: native/packages/react-native-bpk-component-animate-height/readme.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  state = {
+    expanded: true,
+  }
+
+  onToggle= () => {
+    this.setState({ expanded: !this.state.expanded });
+  }
+
+  render() {
+    <View style={styles.container}>
```
[the remark this answers is not in the document]
Comment (georgegillams) [reply native/packages/react-native-bpk-component-animate-height/readme.md]: Good catch!

File: native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js (hunk 5 above)
Comment (georgegillams) [reply native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js]: It complained when I did that!

File: packages/bpk-tokens/src/common/base/animations.json (hunk 7 above)
Comment (georgegillams) [reply packages/bpk-tokens/src/common/base/animations.json]: They are already used on web, and I just added them to the RN tokens too.
This is probably something we need to discuss / get some designer input on

File: native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight-test.common.js (hunk 6 above)
Comment (georgegillams) [reply native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight-test.common.js]: Done 😸

File: native/packages/react-native-bpk-component-animate-height/readme.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    justifyContent: 'center',
+    padding: TOKENS.spacingBase,
+  },
+  animateHeight: {
+    marginBottom: TOKENS.spacingBase,
+  }
+});
+
+export default class App extends Component {
+  state = {
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-animate-height/readme.md]: This is proposed syntax (see https://github.com/tc39/proposal-class-fields). I reckon we should stick to proposals that have actually made the spec (stage 4).

See https://github.com/tc39/proposals for more info.

File: native/packages/react-native-bpk-component-animate-height/readme.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    expanded: true,
+  }
+
+  onToggle= () => {
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-animate-height/readme.md]: So is this.

File: native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+  android: () => require('bpk-tokens/tokens/android/base.react.native.common.js'), // eslint-disable-line global-require
+})();
+
+const {
+  animationDurationBase,
+} = tokens;
+
+class BpkAnimateHeight extends React.Component {
+
+  state = {
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js]: Don't use this syntax :) see previous comments.

File: native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.expanded !== this.state.expanded) {
+      this.state.expanded = nextProps.expanded;
+      this.resize();
+    }
+  }
+
+  setExpandedHeight(event, expandedImmediately) {
+    // if (this.state.expandedHeightSet) return;
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js]: remove comment

File: changelog.md [hunk 13]
```diff
@@ -56,6 +56,7 @@ _Nothing Yet_
 **Added:**
 - bpk-tokens: 26.3.0 => 26.4.0
   - iOS & Android border size tokens
+  - iOS & Android animation duration tokens
```
Comment (georgegillams) [inline changelog.md]: Should be moved to `unreleased`

File: changelog.md [hunk 14]
```diff
@@ -65,6 +66,9 @@ _Nothing Yet_
 - react-native-bpk-component-button: 3.0.5 => 3.1.0
   - Primary and secondary buttons can now be themed.
 
+- react-native-bpk-component-animate-height:
```
Comment (georgegillams) [inline changelog.md]: as above

**Pull request merged**